### ISSUE TITLE
Add reportXXX rule for unreachable code

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#2672]
Pyright flags unreachable code with a `hint` diagnostic. I would like to override this severity using `python.analysis.diagnosticSeverityOverrides`. However, I don't see a corresponding `reportXXX` rule for unreachable code that I see for other diagnostic categories. Can this be added?

**Comment by erictraut** [comment]
This has been discussed before. The tagged hint isn't meant to be displayed like a diagnostic. A few clients incorrectly display it as such. A client can specify whether it supports this particular tagged hint. If it does, it should display them appropriately (as a subtle "dimmed" text). Some clients request this tagged hint and then display it in manner that makes it appear as a normal diagnostic. You should talk to the maintainer of your client and ask them to add a configuration option to disable this tagged hint.

**Comment by smackesey** [comment]
Respectfully, the reason this has been discussed many times before (I have seen it in other issues) is because your stance here is confusing for me and other users who think of PyRight as a language server supporting Language Server Protocol. You seem to be insisting on a definition of "diagnostic" that is different than LSP's definition, despite the fact that PyRight uses LSP. I am sure that issues around hint diagnostics will continue to crop up in other contexts so long as you maintain this stance.

[Diagnostics](https://microsoft.github.io/language-server-protocol/specification#diagnostic) are a specific concept in Language Server Protocol. They can have 1 of 4 severities: `hint`, `info`, `warning`, `error`. The job of the language server is to simply send diagnostics, which Pyright does and does well. Among the many kinds of diagnostics PyRight sends is an `"Code is unreachable"` diagnostic with severity of `hint`.

It is strange to say "the tagged hint isn't meant to be displayed as a diagnostic", because it is literally communicated to the editor a diagnostic. A generic editor-embedded LSP client will simply display it as such, using whatever behavior has been configured to display hint diagnostics. Further, from the perspective of LSP, there is no "correct" behavior for showing hint or any other kind of diagnostic-- that's up to the client. Just because some editors (VSCode) display "hint" diagnostics categorically differently than error/warning diagnostics, doesn't mean every editor needs to do so. Further, there is nowhere in LSP where the server specifies how particular diagnostics are to be displayed. Am I wrong here? Or does PyRight communicate this styling info through some nonstandard structure? If PyRight doesn't communicate this styling info, then on what grounds can you say editors are behaving "incorrectly" when they display the communicated `hint` diagnostic like any other `hint` diagnostic from any other server?

Since PyRight is explicitly advertised as supporting multiple editors, it should not insist on an idiosyncratic definition of "diagnostic" that contradicts the underlying protocol (LSP) it uses to communicate with arbitrary clients.

**Comment by erictraut** [comment]
I think we may be unnecessarily mixing up two different topics.

1. What is the proper interpretation of tagged hints in the LSP, and should pyright provide a separate switch to disable said hints?

2. Should pyright provide a diagnostic check that emits a diagnostic for "unreachable code"?

On point 1, we may need to agree to disagree. The LSP standard provides a way for a client to specify whether it supports specific hint tags. Pyright honors this and does not send any of these tagged hints if the client indicates that it doesn't support (or doesn't want) them. Your client is apparently indicating that it wants to receive hints with this specific tag, so pyright is sending them.

This specific tag is used for several purposes: unreferenced imports, unreferenced input parameters, unreferenced local variables, unreferenced classes, unreferenced functions, and code that is not analyzed because of certain conditions indicated by PEP 484 (e.g. TYPE_CHECKING and sys.version_info comparisons). These are all reasonable use cases for this tagged hint, and pyright is using it in a manner that is consistent with other language servers like the TypeScript LSP.

As I mentioned above, the tagged hint is used for unreferenced symbols. However, these conditions are also covered by separate diagnostic rules: reportUnusedImport, reportUnusedClass, reportUnusedFunction, and reportUnusedVariable. So there is precedent for emitting diagnostics (conditionally) for conditions that also generate tagged hints (unconditionally).

With that context in mind, are you looking for a way to disable tagged hints (point 1), or are you looking for a way to enable diagnostics for unreachable code (point 2)? If it's the former, then I stand by my position that this is something that you should discuss with the maintainer of your client or LSP plugin. If it's the latter, then I'd like to understand your use case in more detail.

**Comment by smackesey** [comment]
Thanks, yes I think your disentanglement of the issues is correct.

On (1): I know that LSP provides [capabilities communication about diagnostic tags] (https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#publishDiagnosticsClientCapabilities). AFAICT there is no concept of "hint tags" specifically-- diagnostic tags can apply to any diagnostic regardless of severity.

If I understand you, your interpretation of diagnostic tags is: "If my server sends a tagged diagnostic, and the client has indicated it doesn't support tags, then my server shouldn't send that diagnostic." IMO this interpretation is not implied by the LSP spec. Instead the LSP spec just implies that you don't need to send the tags if the client doesn't support them-- tag capabilities suggest nothing about whether the diagnostic itself should be sent.

This is further implied by the description of the `Unnecessary` tag: "Clients are allowed to render diagnostics with this tag faded out instead of having an error squiggle." The key word is "allowed". To my ear, tags are intended as additional metadata for "progressive enhancement" of diagnostic display, not as markers to delineate classes of diagnostics that should be toggled on/off based on client tag capabilities.

Rather, the conventional way to control whether a diagnostic is sent is to provide a setting that  lets you toggle on/off (or adjust the severity) of sets of diagnostics according to their `code` property. Though this is not in the LSP spec, almost every language server has some version of this. Pyright's is`python.analysis.diagnosticSeverityOverrides`. IMO this is most simple and straightforward when every kind of diagnostic sent by the language server, regardless of severity, is controllable here via the same API.

The OP is requesting that "Unreachable code" diagnostics (and really IMO all hint diagnostics) should be configurable here. I am a little confused by the phrasing of your question (2), because PyRight *already does provide a diagnostic check that emits a diagnostic for unreachable code*. This is how unreachable code is communicated -- as a diagnostic. I can see it right there in the LSP logs under "diagnostics". A diagnostic with a tag, yes, but still a diagnostic-- matching the `Diagnostic` data structure defined in the LSP spec. But you seem to be using "diagnostic" to mean something else?

AFAICT, you've read something into the LSP spec that isn't there-- namely the idea that there is some kind of fundamental separation between hints and other diagnostics, or tagged diagnostics and other diagnostics, and that this distinction can render items in the `diagnostics` list communicated by the server, conforming to the `Diagnostic` structure defined by the LSP spec, somehow not "diagnostics" but instead "tagged hints". This is a very unintuitive interpretation to me, and I am guessing to many others as well. The only way I can make sense of "tagged hint" is "a diagnostic with severity level hint and one or more diagnostic tags".

I'll rest my case there on point (1), I guess we agree to disagree if you maintain your position.

Now regarding my use case for (2): I've configured my editor to treat hint diagnostics differently than error/warning/info diagnostics. Error/warning/info are treated as action items in a problem list to resolve, hints are grayed out and filtered from my actionable list. I would like to add "Unreachable code" to that actionable list. The natural way to do this is to promote the severity to warning, but because it is only a hint diagnostic and Pyright treats hints differently than all other diagnostics, I can't do that.

**Comment by erictraut** [comment]
For your use case, you've indicated "how" you would like to use this feature, but I'm interested in the "why" and the "what".

I presume that you believe that some (many?) "unreachable code" instances represent potential bugs, and you therefore want to examine (and possibly eliminate) them? Is that a correct assumption?

Pyright currently detects blocks of code that are not reachable for three reasons.
1. Conditional checks that rely on conditions spelled out in PEP 484. This includes conditional statements that take one of the following forms:
   `TYPE_CHECKING`
    `sys.version_info[0] >= X`
    `sys.platform != 'X'`
    `sys.platform == 'X'`
    `os.name != 'X'
    `os.name == 'X'
    `True`
    `False`

![Screen Shot 2021-12-07 at 12 50 55 PM](https://user-images.githubusercontent.com/7040122/145104187-f6987b8b-3b69-4bfc-91f5-c29ee58d883d.png)


2. Code that is provably unreachable based purely on syntax (independent of type analysis). This includes things like code after a `raise` or `return` expression or code after a `while True` loop.

![Screen Shot 2021-12-07 at 12 52 38 PM](https://user-images.githubusercontent.com/7040122/145104413-91af8753-515c-43ae-a98a-c48790d36202.png)


3. Code after a call to a function that returns `NoReturn`

![Screen Shot 2021-12-07 at 12 55 02 PM](https://user-images.githubusercontent.com/7040122/145104713-9734109e-be64-430d-a161-9a6bb2e4a36d.png)


Cases 1 is legitimate and common, not indicative of a bug, so reporting a diagnostic for this case would generate a lot of false positives. Case 2 and 3 are arguably more likely to indicate a bug, but there are legitimate uses for these, such as code that you want to leave in place but temporarily make inaccessible. Most of the time this isn't a bug, but simply some innocuous vestigial code. I'm not sure it should be the job of a type checker to detect this, since it's more of a linting feature and not a type violation.

Pyright currently doesn't report unreachable code based on type analysis. It does detect such code in a few special cases to help eliminate false positives and support exhaustive checks. 

![Screen Shot 2021-12-07 at 1 09 35 PM](https://user-images.githubusercontent.com/7040122/145106605-955593cf-0aae-43e5-a635-fcc15a876bd9.png)

So, I'm looking for some specific examples where you think that detection of "unreachable code" would be useful to expose as a diagnostic. I'd want to be convinced that reporting a diagnostic in these cases is relatively unlikely to result in a false positive because diagnostic checks that report too many false positives are not very useful.

I know that mypy has an option for reporting unreachable code, but in my experience it is quite noisy (lots of false positives) and error prone. It's disabled by default, and I suspect that few people use it.

**Comment by smackesey** [comment]
Thanks for this thorough response.

>I presume that you believe that some (many?) "unreachable code" instances represent potential bugs, and you therefore want to examine (and possibly eliminate) them? Is that a correct assumption?

Yes this is correct. As you surmise, I'd like to treat 2 and 3 like bugs. I understand that others might code in a style where they don't want types 2 and 3 flagged, but I can't remember a time where I've done so-- I will comment out the code if I know it's unused but I want to keep it around. This means 2 and 3 are virtually always bugs for me, and I suspect a significant population of others. This is an argument for letting the user choose.

For a somewhat contrived example of a bug arising from 2, occasionally I have a return statement indented incorrectly, this would catch that:

```python
def outer_function():
    
    def some_longish_inner_function():
        # ... code ...
    return  # oops, this should be indented

    # ... code ...
    return  # this is unreachable bc above return is incorrectly indented
```

**Comment by erictraut** [comment]
OK, thanks for the additional context.

I think we'll pass on this feature for now. I don't think that most users would find it useful. It's not something we could enable by default, even in strict mode. So even for the small number of users who might find it useful, it would go undiscovered in most cases. In general, if a diagnostic check is problematic enough that we cannot enable it by default in "strict" mode, it gives us pause. We would need a strong signal from many users before committing to add and maintain the feature.

We will revisit this decision in the future if this thread receives a significant number of upvotes (thumbs up emoticons).

In my own personal use (and I don't think I'm an outlier), this is not a feature I would ever use because the current functionality (where unreachable code is displayed as "grayed out" in the editor) already alerts me to something that is potentially unexpected, and it does so without the annoyance of false positive errors. The example you provided above, for example, would stand out to me if I was not expecting to see a bunch of grayed-out code at that location.

If we were to implement this feature, I'd need to think about how we might differentiate between cases 1, 2 and 3 above. Currently, they are all handled by the same mechanism (an unreachable node within the code flow graph), and there's no easy way to distinguish between them.

**Comment by quodlibetor** [comment]
I also would like to be able to promote unreachable code to an error -- I ran into this because I just spent a while tracking down a bug that was due to an incorrectly placed return statemt.

In all cases I've ever had intentionally unreachable code I have added comments, and adding a `# pyright: ignore` comment to the beginning of a block that was entirely unreachable would be a significantly smaller burden than spending time debugging items like this.

Making this configurable, and leaving the default level as `hint` seems like a straightforward way to make progress on this.
